src/java/org/apache/cassandra/streaming/StreamSession.java

```diff
@@ -1340,32 +1338,41 @@ public String toString()
                '}';
     }
 
-    public static Throwable boundStackTrace(Throwable e, int limit)
+    public static StringBuilder boundStackTrace(Throwable e, int limit)
     {
         Set<Throwable> visited = Collections.newSetFromMap(new IdentityHashMap<>());
-        return boundStackTrace(e, limit, visited);
+        StringBuilder out = new StringBuilder();
+        return boundStackTrace(e, limit, visited, out);
     }
 
-    public static Throwable boundStackTrace(Throwable e, int limit, Set<Throwable> visited)
+    public static StringBuilder boundStackTrace(Throwable e, int limit, Set<Throwable> visited, StringBuilder out)
     {
         if (e == null)
-            return e;
+            return null;
 
         if (!visited.add(e))
-            return e;
+            return out.append("[CIRCULAR REFERENCE: ").append(e.getClass().getName()).append(": ").append(e.getMessage()).append("]");
         visited.add(e);
 
         if (e.getStackTrace().length == 0 || e.getStackTrace().length < limit)
         {
-            boundStackTrace(e.getCause(), limit, visited);
-            return e;
+            out.append(e.getClass().getName()).append(": ").append(e.getMessage()).append('\n');
+            boundStackTrace(e.getCause(), limit, visited, out);
+            return out;
         }
 
         StackTraceElement[] stackTrace = e.getStackTrace();
-        StackTraceElement[] limitedStackTrace = Arrays.copyOfRange(stackTrace, 0, limit);
-        e.setStackTrace(limitedStackTrace);
-        boundStackTrace(e.getCause(), limit, visited);
+        StringBuilder stackTraceBuilder = new StringBuilder();
+        stackTraceBuilder.append(e.getClass().getName() + ": " + e.getMessage()).append('\n');
+
+        for (int i = 0; i < limit; i++)
+        {
+            stackTraceBuilder.append('\t').append(stackTrace[i]);
+            if (e.getCause() == null && i == limit - 1) continue;
+            stackTraceBuilder.append('\n');
+        }
 
-        return e;
+        boundStackTrace(e.getCause(), limit, visited, out.append(stackTraceBuilder));
```

Comment:
why is `limit` not changed?  if you are only supposed to print 2 stack frames, this logic will do 2 per throwable